### PR TITLE
nsqadmin/nsqd: refactor stats querying

### PR DESCRIPTION
**Pull request by andyxning**
As nsqd `stats` api supports `format`, `topic` and `channel` url parameters. When displaying info for a specific topic and channel, nsqadmin should query nsqd `stats` api with the specified `topic` and `channel` instead querying all channels under the specified `topic` and laterly filter out the specified channel stats. 

Query stats on demand can reduce the time for nsqadmin api to return and will reduce the possibility  that nsqadmin report timeout when to query a topic with many channels and each channel with many consumer clients.

/cc @mreiferson @ploxiln

### CODE REVIEW

File: internal/clusterinfo/data.go [hunk 1]
```diff
@@ -548,7 +549,16 @@ func (c *ClusterInfo) GetNSQDStats(producers Producers, selectedTopic string) ([
 			defer wg.Done()
 
 			addr := p.HTTPAddress()
-			endpoint := fmt.Sprintf("http://%s/stats?format=json", addr)
+
+			var endpoint string
+			if len(selectedChannel) != 0 {
```
Comment (mreiferson) [inline internal/clusterinfo/data.go]: This isn't entirely correct, as it doesn't test for `selectedTopic`, let's just adopt the code in #925 wholesale for this block:

```go
if selectedTopic == "" {
	endpoint = fmt.Sprintf("http://%s/stats?format=json", addr)
} else if selectChannel == "" {
	endpoint = fmt.Sprintf("http://%s/stats?topic=%s&format=json", addr, selectedTopic)
} else {
	endpoint = fmt.Sprintf("http://%s/stats?topic=%s&channel=%s&format=json", addr, selectedTopic, selectChannel)
}
```

File: nsqd/stats.go [hunk 2]
```diff
@@ -113,20 +113,32 @@ type ChannelsByName struct {
...
 func (c ChannelsByName) Less(i, j int) bool { return c.Channels[i].name < c.Channels[j].name }
 
-func (n *NSQD) GetStats() []TopicStats {
+func (n *NSQD) GetStats(topic string, channel string) []TopicStats {
 	n.RLock()
-	realTopics := make([]*Topic, 0, len(n.topicMap))
-	for _, t := range n.topicMap {
-		realTopics = append(realTopics, t)
+	var realTopics []*Topic
+	if len(topic) == 0 {
```
Comment (mreiferson) [inline nsqd/stats.go]: Minor, but In Go, this test should just be:

```go
if topic == "" {
```

File: nsqd/stats.go [hunk 3]
```diff
@@ -113,20 +113,32 @@ type ChannelsByName struct {
...
+		realTopics = make([]*Topic, 0, len(n.topicMap))
+		for _, t := range n.topicMap {
+			realTopics = append(realTopics, t)
+		}
+	} else {
+		realTopics = make([]*Topic, 0, 1)
+		realTopics = append(realTopics, n.topicMap[topic])
```
Comment (mreiferson) [inline nsqd/stats.go]: These two lines can be simplified to:

```go
realTopics = []*Topic{n.topicMap[topic]}
```

File: nsqd/stats.go [hunk 4]
```diff
@@ -113,20 +113,32 @@ type ChannelsByName struct {
...
 	n.RUnlock()
 	sort.Sort(TopicsByName{realTopics})
 	topics := make([]TopicStats, 0, len(realTopics))
 	for _, t := range realTopics {
 		t.RLock()
-		realChannels := make([]*Channel, 0, len(t.channelMap))
-		for _, c := range t.channelMap {
-			realChannels = append(realChannels, c)
+		var realChannels []*Channel
+		if len(channel) == 0 {
```
Comment (mreiferson) [inline nsqd/stats.go]: Same idiomatic improvement here:

```go
if channel == "" {
```

File: nsqd/stats.go [hunk 5]
```diff
@@ -113,20 +113,32 @@ type ChannelsByName struct {
...
+			realChannels = make([]*Channel, 0, len(t.channelMap))
+			for _, c := range t.channelMap {
+				realChannels = append(realChannels, c)
+			}
+		} else {
+			realChannels = make([]*Channel, 0, 1)
+			realChannels = append(realChannels, t.channelMap[channel])
```
Comment (mreiferson) [inline nsqd/stats.go]: And same simplification here:

```go
realChannels = []*Channel{t.channelMap[channel]}
```

File: internal/clusterinfo/data.go (hunk 1 above)
Comment (ploxiln) [reply internal/clusterinfo/data.go]: Let me humbly offer my own variant here:

```go
endpoint := fmt.Sprintf("http://%s/stats?format=json", addr)
if selectedTopic != "" {
        endpoint += "&topic=" + selectedTopic
        if selectedChannel != "" {
                endpoint += "&channel=" + selectedChannel
        }
}
```

Comment (mreiferson) [reply internal/clusterinfo/data.go]: 👍

File: nsqd/stats.go (hunk 2 above)
Comment (andyxning) [reply nsqd/stats.go]: Will do.

File: nsqd/stats.go (hunk 3 above)
Comment (andyxning) [reply nsqd/stats.go]: Will do.

File: nsqd/stats.go (hunk 4 above)
Comment (andyxning) [reply nsqd/stats.go]: Will do.

File: nsqd/stats.go (hunk 5 above)
Comment (andyxning) [reply nsqd/stats.go]: Will do.

File: internal/clusterinfo/data.go (hunk 1 above)
Comment (andyxning) [reply internal/clusterinfo/data.go]: Thanks for your review and the more simplified and clearer refactor.

**Comment by andyxning** [comment]
@mreiferson @ploxiln Comments addressed. PTAL.

File: nsqd/stats_test.go [hunk 6]
```diff
@@ -32,7 +32,7 @@ func TestStats(t *testing.T) {
 	identify(t, conn, nil, frameTypeResponse)
 	sub(t, conn, topicName, "ch")
 
-	stats := nsqd.GetStats()
+	stats := nsqd.GetStats(topicName, "ch")
```
Comment (ploxiln) [inline nsqd/stats_test.go]: it's probably a good idea to expand this test a bit - put a message to one other topic and channel

**Comment by ploxiln** [comment]
looking pretty good to me

File: nsqd/stats_test.go (hunk 6 above)
Comment (andyxning) [reply nsqd/stats_test.go]: Sounds reasonable. Done.

**Comment by andyxning** [comment]
@ploxiln Comments addressed. PTAL.

File: nsqd/stats.go [hunk 7]
```diff
@@ -113,20 +113,30 @@ type ChannelsByName struct {
...
-	for _, t := range n.topicMap {
-		realTopics = append(realTopics, t)
+	var realTopics []*Topic
+	if topic == "" {
+		realTopics = make([]*Topic, 0, len(n.topicMap))
+		for _, t := range n.topicMap {
+			realTopics = append(realTopics, t)
+		}
+	} else {
+		realTopics = []*Topic{n.topicMap[topic]}
```
Comment (mreiferson) [inline nsqd/stats.go]: Just realized we're not checking if the topic exists in `topicMap`, can you add that check here?

File: nsqd/stats.go [hunk 8]
```diff
@@ -113,20 +113,30 @@ type ChannelsByName struct {
...
-		for _, c := range t.channelMap {
-			realChannels = append(realChannels, c)
+		var realChannels []*Channel
+		if channel == "" {
+			realChannels = make([]*Channel, 0, len(t.channelMap))
+			for _, c := range t.channelMap {
+				realChannels = append(realChannels, c)
+			}
+		} else {
+			realChannels = []*Channel{t.channelMap[channel]}
```
Comment (mreiferson) [inline nsqd/stats.go]: same here, need to check if channel is in `channelMap` first

File: nsqd/stats.go (hunk 7 above)
Comment (andyxning) [reply nsqd/stats.go]: @mreiferson Good catch! I also have enhanced the test for both of none exist topic and channel.

PTAL.

**Comment by mreiferson** [comment]
@andyxning fixed up the commits, LGTM, thanks!
